content/blog/2019-11-13-DynamicEdgeProfiling/index.md

```diff
@@ -0,0 +1,122 @@
+
+## Algorithm
+To select the edges to profile, we used Knuth's algorithm:
+
+---
+<!-- ![](https://i.imgur.com/NbavLnq.png) -->
+<img src="array.png" style="max-width: 100%" >
+---
+Essentially, this algorithm first finds the (acyclic) minimum spanning tree of the edges in a function and then adds profiling instrumentation---code that logs profiling information---to all edges not in the spanning tree. Note that this algorithm uses the first block in a function as the root of the spanning tree, and that edges are treated as bidirectional when checking for cycles.
+
+This algorithm works because the acyclic nature of the spanning tree entails that there is exactly one path between any two instrumented edges (or the entry or exit of the function). Thus, logging the traversal of one instrumented edge implies traversal of the program through the path from the previously logged edge location to the current one. This also implies that this is the minimum number of instrumented edges because having one less instrumented edge would either create a path that could be traversed without logging any profiling information or create multiple paths that lack sufficient profiling information to deduce which path was taken. 
+
+### Offline Edge Count Extrapolation Algorithm
+After we run our program, we know the number of times our instrumented edges were traversed. In order to determine the number of times the remaining edges were produced, we need to do some post-processing of our CFG with our runtime results.
+
+The idea behind this algorithm is to iterate over the edges and fill in the edge count until convergence. The program will converge under the assumption that the edge profiling information supplied is sufficient to determine the rest of the edge counts. It is valid to fill in an edge count if there is a simple path between two nodes with known edge count. You can find detailed pseudocode and proofs for this offline algorithm in the [LLVM Publication](https://llvm.org/pubs/2010-04-NeustifterProfiling.pdf).
+
+## Implementation
+
+We wanted our instrumentation implementation to be as noninvasive as possible for the base code as pervasive changes could make the LLVM IR harder to reason about and could prevent future optimizations. To achieve that, we implemented the bulk of the profiling information logging in a separate runtime library file. This strategy allows us to reduce instrumentation to two function calls, one to `logsrc(num1)` in the edge's "source" block and one to `logdest(num2)` in the "destination" block, both of which reside in the library file. These functions calls take in a block-unique integer (`num1` and `num2` above) that is generated during the pass as arguments, meaning that we can identify an edge traversal with a pair of `logsrc(num1), logdest(num2)` function calls.
```

Comment:
Just for fun, it would be cool to know how you determined that block-unique integer. Perhaps just the position in the function's block list?